### PR TITLE
MVP unified MCMC

### PR DESCRIPTION
**Pull request by ieva-improbable**
An MVP implementation for unified MCMC - a class that picks the right sampling algorithm for you given a Bayesian network ([SOL-2428](https://improbableio.atlassian.net/browse/SOL-2428)). NUTS will be chosen if the network is differentiable while Metropolis Hastings will be chosen if some vertex in the network is not differentiable.

Usage:
```java
 PosteriorSamplingAlgorithm samplingAlgorithm = MCMC.withDefaultConfig().forNetwork(yourNetwork);
 NetworkSamples samples = samplingAlgorithm.getPosteriorSamples(...);
```

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/UnifiedMCMCTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class UnifiedMCMCTest {
+
+    @Test
+    public void checksIfNetworkIsDifferentiable() {
+        BayesianNetwork network = mock(BayesianNetwork.class);
+        Vertex vertex = new GaussianVertex(0, 1);
+        when(network.getTopLevelLatentVertices()).thenReturn(ImmutableList.of(vertex));
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/UnifiedMCMCTest.java]: Do you still need to mock here? Can you just create a BayesianNetwork from the single Gaussian?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/UnifiedMCMCTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        when(network.getLatentOrObservedVertices()).thenReturn(ImmutableList.of(vertex));
+        when(network.getContinuousLatentVertices()).thenReturn(ImmutableList.of(vertex));
+
+        MCMC.withDefaultConfig().forNetwork(network);
+        verify(network).getNonDifferentiableVertices();
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/UnifiedMCMCTest.java]: Oh I see why you need the mock. Is it a requirement that the unit under test calls this method? Or is it's only requirement that it returns the appropriate sampling algorithm?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTS.java [hunk 3]
```diff
@@ -77,6 +78,7 @@ public NetworkSamples getPosteriorSamples(final BayesianNetwork bayesNet,
                                               final List<? extends Vertex> sampleFromVertices,
                                               final int sampleCount) {
 
+        Preconditions.checkArgument(!sampleFromVertices.isEmpty(), "List of vertices to sample from is empty");
```
Comment (icantsayrural) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTS.java]: Just curious, why was this added now?

Comment (ieva-improbable) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTS.java]: We were doing some mocking for testing and the realised that NUTS doesn't terminate if you pass in an empty list of vertices. So added the check in case someone else runs into this issue.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/UnifiedMCMCTest.java (hunk 2 above)
Comment (ieva-improbable) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/UnifiedMCMCTest.java]: I guess the latter, will remove that particular test.
